File: src/lib/geodesy.ts
import contractJson from "../data/geodesy-contract.json";

export const GEODESY_CONTRACT_VERSION = contractJson.contract_version;
export const WGS84_A_M = 6_378_137;
export const WGS84_INV_F = 298.257_223_563;

export type VerticalDatum =
  | "wgs84_ellipsoid"
  | "navd88_geoid18"
  | "idealized_mean_sea_level"
  | "depth_below_idealized_mean_sea_level"
  | "local_enu";
export type VerticalAxis = "positive_up" | "positive_down";

export type HeightFieldMetadata = {
  horizontal_crs: string;
  vertical_datum: VerticalDatum;
  vertical_axis: VerticalAxis;
  unit: "metre";
  declared_vertical_error_m: number;
};

export const IDEALIZED_SEA_SURFACE_HEIGHT_FIELD: HeightFieldMetadata = {
  horizontal_crs: "EPSG:4326",
  vertical_datum: "idealized_mean_sea_level",
  vertical_axis: "positive_up",
  unit: "metre",
  declared_vertical_error_m: 4000,
};

export const COARSE_BATHYMETRY_DEPTH_FIELD: HeightFieldMetadata = {
  horizontal_crs: "EPSG:4326",
  vertical_datum: "depth_below_idealized_mean_sea_level",
  vertical_axis: "positive_down",
  unit: "metre",
  declared_vertical_error_m: 4000,
};

export type GeodeticPosition = {
  latDeg: number;
  lonDeg: number;
  ellipsoidHeightM: number;
};

export type EcefPosition = { xM: number; yM: number; zM: number };
export type LocalEnu = { eastM: number; northM: number; upM: number };
export type UnrealPositionCm = { xEastCm: number; yNorthCm: number; zUpCm: number };

export function convertVerticalHeight(
  valueM: number,
  from: VerticalDatum,
  to: VerticalDatum,
  context: { geoidUndulationM?: number } = {},
): number {
  if (!Number.isFinite(valueM)) throw new Error("Vertical value must be finite.");
  if (from === to) return valueM;
  if (from === "wgs84_ellipsoid" && to === "navd88_geoid18") {
    if (!Number.isFinite(context.geoidUndulationM)) {
      throw new Error("WGS84 ellipsoid to NAVD88 requires a reviewed geoid undulation.");
    }
    return valueM - (context.geoidUndulationM as number);
  }
  if (from === "navd88_geoid18" && to === "wgs84_ellipsoid") {
    if (!Number.isFinite(context.geoidUndulationM)) {
      throw new Error("NAVD88 to WGS84 ellipsoid requires a reviewed geoid undulation.");
    }
    return valueM + (context.geoidUndulationM as number);
  }
  if (
    (from === "idealized_mean_sea_level" && to === "depth_below_idealized_mean_sea_level") ||
    (from === "depth_below_idealized_mean_sea_level" && to === "idealized_mean_sea_level")
  ) {
    return -valueM;
  }
  throw new Error(`Unsupported vertical datum conversion: ${from} to ${to}.`);
}

export function geodeticToEcef(position: GeodeticPosition): EcefPosition {
  if (
    !Number.isFinite(position.latDeg) ||
    !Number.isFinite(position.lonDeg) ||
    !Number.isFinite(position.ellipsoidHeightM) ||
    Math.abs(position.latDeg) > 90 ||
    Math.abs(position.lonDeg) > 180
  ) {
    throw new Error("Geodetic coordinates must be finite WGS84 longitude/latitude/height values.");
  }
  const flattening = 1 / WGS84_INV_F;
  const eccentricitySq = flattening * (2 - flattening);
  const lat = position.latDeg * Math.PI / 180;
  const lon = position.lonDeg * Math.PI / 180;
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);
  const primeVertical = WGS84_A_M / Math.sqrt(1 - eccentricitySq * sinLat * sinLat);
  return {
    xM: (primeVertical + position.ellipsoidHeightM) * cosLat * Math.cos(lon),
    yM: (primeVertical + position.ellipsoidHeightM) * cosLat * Math.sin(lon),
    zM: (primeVertical * (1 - eccentricitySq) + position.ellipsoidHeightM) * sinLat,
  };
}

export function ecefToLocalEnu(point: EcefPosition, origin: GeodeticPosition): LocalEnu {
  const base = geodeticToEcef(origin);
  const lat = origin.latDeg * Math.PI / 180;
  const lon = origin.lonDeg * Math.PI / 180;
  const dx = point.xM - base.xM;
  const dy = point.yM - base.yM;
  const dz = point.zM - base.zM;
  return {
    eastM: -Math.sin(lon) * dx + Math.cos(lon) * dy,
    northM: -Math.sin(lat) * Math.cos(lon) * dx - Math.sin(lat) * Math.sin(lon) * dy + Math.cos(lat) * dz,
    upM: Math.cos(lat) * Math.cos(lon) * dx + Math.cos(lat) * Math.sin(lon) * dy + Math.sin(lat) * dz,
  };
}

export function enuToUnrealCm(position: LocalEnu): UnrealPositionCm {
  return {
    xEastCm: position.eastM * 100,
    yNorthCm: position.northM * 100,
    zUpCm: position.upM * 100,
  };
}

export function getGeodesyDiagnosticsSnapshot() {
  return structuredClone(contractJson);
}

File: src-tauri/src/data/geodesy.rs
//! Renderer-neutral coordinate and vertical-datum contract.
//!
//! Cesium and Rust exchange WGS 84 geodetic/ECEF coordinates in metres.
//! Unreal consumes a local east/north/up frame in centimetres. Vertical
//! conversions are deliberately explicit: no global geoid or tide grid is
//! bundled, so conversions that lack the required correction fail closed.

use serde::{Deserialize, Serialize};

pub const WGS84_A_M: f64 = 6_378_137.0;
pub const WGS84_INV_F: f64 = 298.257_223_563;
pub const HORIZONTAL_CRS_GEOGRAPHIC_2D: &str = "EPSG:4326";
pub const HORIZONTAL_CRS_GEOGRAPHIC_3D: &str = "EPSG:4979";
pub const HORIZONTAL_CRS_ECEF: &str = "EPSG:4978";
pub const CONTRACT_VERSION: &str = "1.0.0";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VerticalDatum {
    Wgs84Ellipsoid,
    Navd88Geoid18,
    IdealizedMeanSeaLevel,
    DepthBelowIdealizedMeanSeaLevel,
    LocalEnu,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VerticalAxis {
    PositiveUp,
    PositiveDown,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HeightFieldMetadata {
    pub horizontal_crs: String,
    pub vertical_datum: VerticalDatum,
    pub vertical_axis: VerticalAxis,
    pub unit: String,
    pub declared_vertical_error_m: f64,
}

pub fn sea_surface_height_field() -> HeightFieldMetadata {
    HeightFieldMetadata {
        horizontal_crs: HORIZONTAL_CRS_GEOGRAPHIC_2D.to_string(),
        vertical_datum: VerticalDatum::IdealizedMeanSeaLevel,
        vertical_axis: VerticalAxis::PositiveUp,
        unit: "metre".to_string(),
        declared_vertical_error_m: 4_000.0,
    }
}

pub fn bathymetry_depth_field() -> HeightFieldMetadata {
    HeightFieldMetadata {
        horizontal_crs: HORIZONTAL_CRS_GEOGRAPHIC_2D.to_string(),
        vertical_datum: VerticalDatum::DepthBelowIdealizedMeanSeaLevel,
        vertical_axis: VerticalAxis::PositiveDown,
        unit: "metre".to_string(),
        declared_vertical_error_m: 4_000.0,
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct GeodeticPosition {
    pub lat_deg: f64,
    pub lon_deg: f64,
    pub ellipsoid_height_m: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct EcefPosition {
    pub x_m: f64,
    pub y_m: f64,
    pub z_m: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct LocalEnu {
    pub east_m: f64,
    pub north_m: f64,
    pub up_m: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct UnrealPositionCm {
    pub x_east_cm: f64,
    pub y_north_cm: f64,
    pub z_up_cm: f64,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct VerticalConversionContext {
    /// Geoid undulation N in h = H + N. Required for WGS84 ellipsoid ↔
    /// NAVD88/GEOID18 fixture conversion.
    pub geoid_undulation_m: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VerticalConversionError {
    MissingGeoidUndulation,
    UnsupportedDatumPair,
    NonFiniteValue,
}

pub fn convert_vertical(
    value_m: f64,
    from: VerticalDatum,
    to: VerticalDatum,
    context: VerticalConversionContext,
) -> Result<f64, VerticalConversionError> {
    if !value_m.is_finite() {
        return Err(VerticalConversionError::NonFiniteValue);
    }
    if from == to {
        return Ok(value_m);
    }
    match (from, to) {
        (VerticalDatum::Wgs84Ellipsoid, VerticalDatum::Navd88Geoid18) => context
            .geoid_undulation_m
            .filter(|value| value.is_finite())
            .map(|undulation| value_m - undulation)
            .ok_or(VerticalConversionError::MissingGeoidUndulation),
        (VerticalDatum::Navd88Geoid18, VerticalDatum::Wgs84Ellipsoid) => context
            .geoid_undulation_m
            .filter(|value| value.is_finite())
            .map(|undulation| value_m + undulation)
            .ok_or(VerticalConversionError::MissingGeoidUndulation),
        (VerticalDatum::IdealizedMeanSeaLevel, VerticalDatum::DepthBelowIdealizedMeanSeaLevel)
        | (VerticalDatum::DepthBelowIdealizedMeanSeaLevel, VerticalDatum::IdealizedMeanSeaLevel) => {
            Ok(-value_m)
        }
        _ => Err(VerticalConversionError::UnsupportedDatumPair),
    }
}

pub fn geodetic_to_ecef(position: GeodeticPosition) -> Option<EcefPosition> {
    if !position.lat_deg.is_finite()
        || !position.lon_deg.is_finite()
        || !position.ellipsoid_height_m.is_finite()
        || !(-90.0..=90.0).contains(&position.lat_deg)
        || !(-180.0..=180.0).contains(&position.lon_deg)
    {
        return None;
    }
    let flattening = 1.0 / WGS84_INV_F;
    let eccentricity_sq = flattening * (2.0 - flattening);
    let lat = position.lat_deg.to_radians();
    let lon = position.lon_deg.to_radians();
    let sin_lat = lat.sin();
    let cos_lat = lat.cos();
    let prime_vertical = WGS84_A_M / (1.0 - eccentricity_sq * sin_lat * sin_lat).sqrt();
    Some(EcefPosition {
        x_m: (prime_vertical + position.ellipsoid_height_m) * cos_lat * lon.cos(),
        y_m: (prime_vertical + position.ellipsoid_height_m) * cos_lat * lon.sin(),
        z_m: (prime_vertical * (1.0 - eccentricity_sq) + position.ellipsoid_height_m) * sin_lat,
    })
}

pub fn ecef_to_local_enu(point: EcefPosition, origin: GeodeticPosition) -> Option<LocalEnu> {
    let origin_ecef = geodetic_to_ecef(origin)?;
    let lat = origin.lat_deg.to_radians();
    let lon = origin.lon_deg.to_radians();
    let dx = point.x_m - origin_ecef.x_m;
    let dy = point.y_m - origin_ecef.y_m;
    let dz = point.z_m - origin_ecef.z_m;
    Some(LocalEnu {
        east_m: -lon.sin() * dx + lon.cos() * dy,
        north_m: -lat.sin() * lon.cos() * dx - lat.sin() * lon.sin() * dy + lat.cos() * dz,
        up_m: lat.cos() * lon.cos() * dx + lat.cos() * lon.sin() * dy + lat.sin() * dz,
    })
}

pub fn enu_to_unreal_cm(position: LocalEnu) -> UnrealPositionCm {
    UnrealPositionCm {
        x_east_cm: position.east_m * 100.0,
        y_north_cm: position.north_m * 100.0,
        z_up_cm: position.up_m * 100.0,
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct GeodesyDiagnostics {
    pub contract_version: &'static str,
    pub geographic_2d_crs: &'static str,
    pub geographic_3d_crs: &'static str,
    pub ecef_crs: &'static str,
    pub cesium_frame: &'static str,
    pub unreal_frame: &'static str,
    pub arbitrary_geoid_conversion: &'static str,
}

pub fn diagnostics() -> GeodesyDiagnostics {
    GeodesyDiagnostics {
        contract_version: CONTRACT_VERSION,
        geographic_2d_crs: HORIZONTAL_CRS_GEOGRAPHIC_2D,
        geographic_3d_crs: HORIZONTAL_CRS_GEOGRAPHIC_3D,
        ecef_crs: HORIZONTAL_CRS_ECEF,
        cesium_frame: "EPSG:4978 ECEF metres",
        unreal_frame: "local ENU centimetres: X east, Y north, Z up",
        arbitrary_geoid_conversion: "fail-closed without a supplied geoid undulation",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Contract {
        error_budget: ErrorBudget,
        coastal_benchmarks: Vec<Benchmark>,
    }

    #[derive(Deserialize)]
    struct ErrorBudget {
        geodetic_to_ecef_m: f64,
        ecef_to_local_enu_m: f64,
        fixture_vertical_conversion_m: f64,
    }

    #[derive(Deserialize)]
    struct Benchmark {
        lat_deg: f64,
        lon_deg: f64,
        orthometric_height_m: f64,
        geoid_undulation_m: f64,
        ellipsoid_height_m: f64,
        expected_ecef_m: [f64; 3],
        expected_unreal_local_cm: [f64; 3],
    }

    fn contract() -> Contract {
        serde_json::from_str(include_str!("../../../src/data/geodesy-contract.json"))
            .expect("geodesy contract fixture must parse")
    }

    #[test]
    fn coastal_fixture_vertical_conversions_are_explicit_and_reversible() {
        let contract = contract();
        assert_eq!(contract.coastal_benchmarks.len(), 3);
        for fixture in contract.coastal_benchmarks {
            let context = VerticalConversionContext {
                geoid_undulation_m: Some(fixture.geoid_undulation_m),
            };
            let ellipsoid = convert_vertical(
                fixture.orthometric_height_m,
                VerticalDatum::Navd88Geoid18,
                VerticalDatum::Wgs84Ellipsoid,
                context,
            )
            .unwrap();
            assert!(
                (ellipsoid - fixture.ellipsoid_height_m).abs()
                    <= contract.error_budget.fixture_vertical_conversion_m
            );
            let restored = convert_vertical(
                ellipsoid,
                VerticalDatum::Wgs84Ellipsoid,
                VerticalDatum::Navd88Geoid18,
                context,
            )
            .unwrap();
            assert!(
                (restored - fixture.orthometric_height_m).abs()
                    <= contract.error_budget.fixture_vertical_conversion_m
            );
        }
    }

    #[test]
    fn cesium_ecef_and_unreal_local_frames_match_contract_fixtures() {
        let contract = contract();
        for fixture in contract.coastal_benchmarks {
            let position = GeodeticPosition {
                lat_deg: fixture.lat_deg,
                lon_deg: fixture.lon_deg,
                ellipsoid_height_m: fixture.ellipsoid_height_m,
            };
            let ecef = geodetic_to_ecef(position).unwrap();
            for (actual, expected) in [ecef.x_m, ecef.y_m, ecef.z_m]
                .into_iter()
                .zip(fixture.expected_ecef_m)
            {
                assert!((actual - expected).abs() <= contract.error_budget.geodetic_to_ecef_m);
            }
            let origin = GeodeticPosition {
                ellipsoid_height_m: 0.0,
                ..position
            };
            let unreal = enu_to_unreal_cm(ecef_to_local_enu(ecef, origin).unwrap());
            for (actual, expected) in [unreal.x_east_cm, unreal.y_north_cm, unreal.z_up_cm]
                .into_iter()
                .zip(fixture.expected_unreal_local_cm)
            {
                assert!(
                    (actual - expected).abs() <= contract.error_budget.ecef_to_local_enu_m * 100.0
                );
            }
        }
    }

    #[test]
    fn arbitrary_datum_conversion_fails_closed_without_a_model_value() {
        assert_eq!(
            convert_vertical(
                0.0,
                VerticalDatum::Wgs84Ellipsoid,
                VerticalDatum::Navd88Geoid18,
                VerticalConversionContext::default(),
            ),
            Err(VerticalConversionError::MissingGeoidUndulation),
        );
        assert_eq!(
            convert_vertical(
                0.0,
                VerticalDatum::IdealizedMeanSeaLevel,
                VerticalDatum::Wgs84Ellipsoid,
                VerticalConversionContext::default(),
            ),
            Err(VerticalConversionError::UnsupportedDatumPair),
        );
    }
}
